File: homework01/vigenere.py
def encrypt_vigenere(plaintext: str, keyword: str) -> str:
    """
    Encrypts plaintext using a Vigenere cipher.

    >>> encrypt_vigenere("PYTHON", "A")
    'PYTHON'
    >>> encrypt_vigenere("python", "a")
    'python'
    >>> encrypt_vigenere("ATTACKATDAWN", "LEMON")
    'LXFOPVEFRNHR'
    """
    ciphertext = ""
    a_first = ord("a")
    z_last = ord("z")
    A_first = ord("A")
    Z_last = ord("Z")
    language = 26
    if int(len(keyword)) < int(len(plaintext)):
        keyword = keyword * int(len(plaintext))
        keyword = keyword[: len(plaintext)]
    for i, j in zip(keyword, plaintext):
        code_letter = ord(j)
        code_letter_keyword = ord(i)
        if a_first <= code_letter_keyword <= z_last:
            shift = code_letter_keyword - a_first
        elif A_first <= code_letter_keyword <= Z_last:
            shift = code_letter_keyword - A_first
        if a_first <= code_letter <= z_last:
            k = chr((code_letter - a_first + shift) % language + a_first)
            ciphertext += k
        elif A_first <= code_letter <= Z_last:
            k = chr((code_letter - A_first + shift) % language + A_first)
            ciphertext += k
        else:
            ciphertext += j

    return ciphertext


def decrypt_vigenere(ciphertext: str, keyword: str) -> str:
    """
    Decrypts a ciphertext using a Vigenere cipher.

    >>> decrypt_vigenere("PYTHON", "A")
    'PYTHON'
    >>> decrypt_vigenere("python", "a")
    'python'
    >>> decrypt_vigenere("LXFOPVEFRNHR", "LEMON")
    'ATTACKATDAWN'
    """
    plaintext = ""
    a_first = ord("a")
    z_last = ord("z")
    A_first = ord("A")
    Z_last = ord("Z")
    language = 26
    if int(len(keyword)) < int(len(ciphertext)):
        keyword = keyword * int(len(ciphertext))
        keyword = keyword[: len(ciphertext)]
    for i, j in zip(keyword, ciphertext):
        code_letter = ord(j)
        code_letter_keyword = ord(i)
        if a_first <= code_letter_keyword <= z_last:
            shift = code_letter_keyword - a_first
        elif A_first <= code_letter_keyword <= Z_last:
            shift = code_letter_keyword - A_first
        if a_first <= code_letter <= z_last:
            check_negative = code_letter - a_first - shift
            if check_negative < 0:
                check_negative = (language - (check_negative * (-1) % language)) % language
            k = chr(check_negative + a_first)
            plaintext += k
        elif A_first <= code_letter <= Z_last:
            check_negative = code_letter - A_first - shift
            if check_negative < 0:
                check_negative = (language - (check_negative * (-1) % language)) % language
            k = chr(check_negative + A_first)
            plaintext += k
        else:
            plaintext += j

    return plaintext

File: homework01-go/vigenere/vigenere.go
package vigenere

import "strings"

func EncryptVigenere(plaintext string, keyword string) string {
	var ciphertext string
	var language int = 26

	razn := len(plaintext) - len(keyword)
	k_word := len(keyword)
	for i := 0; i < razn; i++ {
		keyword = keyword + string(keyword[i%k_word])
	}
	keyword = strings.ToLower(keyword)

	for i := 0; i < len(plaintext); i++ {
		shift := int(keyword[0]) - int('a')
		keyword = keyword[1:]
		sym_num := int(plaintext[i])
		if (int('A') <= sym_num) && (sym_num <= int('Z')) {
			sym_num = (int(sym_num)+shift-int('A'))%language + int('A')
		} else if (int('a') <= sym_num) && (sym_num <= int('z')) {
			sym_num = (int(sym_num)+shift-int('a'))%language + int('a')
		}
		ciphertext += string(rune(sym_num))
	}
	return ciphertext
}

func DecryptVigenere(ciphertext string, keyword string) string {
	var plaintext string
	var language int = 26

	razn := len(ciphertext) - len(keyword)
	k_word := len(keyword)
	for i := 0; i < razn; i++ {
		keyword = keyword + string(keyword[i%k_word])
	}
	keyword = strings.ToLower(keyword)

	for i := 0; i < len(ciphertext); i++ {
		shift := int(keyword[0]) - int('a')
		keyword = keyword[1:]
		sym_num := int(ciphertext[i])
		if (int('A') <= sym_num) && (sym_num <= int('Z')) {
			sym_num = (sym_num-shift-int('A')+language)%language + int('A')
		} else if (int('a') <= sym_num) && (sym_num <= int('z')) {
			sym_num = (sym_num-shift-int('a')+language)%language + int('a')
		}
		plaintext += string(rune(sym_num))
	}
	return plaintext
}
